### DISCUSSION TITLE
using .netmiko.yml for inventory

### DISCUSSION THREAD

**Discussion by mikenugs** [ktbyers/netmiko#3226] [Q&A]
Hello,
 
I am looking into creating a .netmiko.yml inventory file for scripting. It appears that using things like netmiko-grep use this by default but I do not see any examples where people have used this for writing their own scripts. Does anyone have any examples where they have created a .netmiko.yml inventory file and used it to either import a single device or a group of devices?

     Thanks,

        Mike F.

**Comment by ktbyers** [comment]
@mikenugs I talk about the inventory format here:

https://pynet.twb-tech.com/blog/netmiko-grep-command-line-utility.html

Let me know if that provides you with enough information.

**Reply by mikenugs** [reply]
That doc is perfect for creating the inventory file, but when I created a script using netmiko that was not related to netmiko-grep it did not automatically look for the .netmiko.yml file. I would like to create the .netmiko.yml to use with things like netmiko-grep but also use it with regular scripts that I write that would ideally find the host information from the .netmiko.yml file.

**Reply by ktbyers** [reply]
Probably can use:

```
from netmiko.utilities import load_devices
```

Then use that function to load the devices and groups.

You would have to add some logic for how to resulting data structure (i.e. devices and groups of devices).

**Reply by mikenugs** [reply]
That worked out perfectly just two more small questions. 

1.) When I create a grouping of devices I am setting the name of the first entry in the list to test_distro for example and when I display this output after importing the .netmiko.yml file it displays the entry for test_distro as a string and not the dictionary that I had specified in the file.

```
---
test_distro:
    device_type: cisco_ios
    host: 10.0.5.5
    username: user
    password: password
test_access:
    device_type: cisco_nxos
    host: 10.0.6.6
    username: user
    password: password

DEVICES:
    - test_distro
    - test_access
```
```
{'DEVICES': ['test_distro', 'test_access'],
 'test_access': {'device_type': 'cisco_nxos',
                 'host': '10.0.6.6',
                 'password': 'password',
                 'username': 'user'},
 'test_distro': {'device_type': 'cisco_ios',
                 'host': '10.0.5.5',
                 'password': 'password',
                 'username': 'user'}}
```

2.) How could I go about using getpass() as the password using .netmiko.yml as apposed to manually specifying the password entry for each device that I define?

**Reply by ktbyers** [reply]
In this:

```python
{'DEVICES': ['test_distro', 'test_access'],
 'test_access': {'device_type': 'cisco_nxos',
                 'host': '10.0.6.6',
                 'password': 'password',
                 'username': 'user'},
 'test_distro': {'device_type': 'cisco_ios',
                 'host': '10.0.5.5',
                 'password': 'password',
                 'username': 'user'}}
```

You can differentiate between groups and devices using the Python data type of the value.

So here there are three keys in the outermost dictionary: "DEVICES", "test_access", "test_distro". And the Python type for the values of each of those three keys is List, Dictionary, Dictionary.

The first one ("DEVICES" being a List, tells you that it is a "group" of devices and the entries in the list will be the keys that belong to that group (so both "test_access" and "test_distro" belong to the DEVICES group).

So this is a List (and thus a group):

```python
In [4]: netmiko_dev["DEVICES"]
Out[4]: ['test_distro', 'test_access']
```

And you could retrieve the device dictionaries just by looking them up:

```python
In [10]: for device in netmiko_dev["DEVICES"]:
    ...:     print(netmiko_dev[device])
    ...: 
{
    'device_type': 'cisco_ios',
    'host': '10.0.5.5',
    'password': 'password',
    'username': 'user'
}
{
    'device_type': 'cisco_nxos',
    'host': '10.0.6.6',
    'password': 'password',
    'username': 'user'
}
```

On your second question, remove the "password" and/or "secret" field and then prompt for the password using `getpass()` and loop over the devices and set the "password" and/or "secret" key.

```python
In [10]: for name, group_or_device in netmiko_dev.items():
    ...:     # Skip the groups for now
    ...:     if isinstance(group_or_device, list):
    ...:         continue
    ...:     group_or_device["password"] = password
    ...:     print(group_or_device)
    ...: 
{
    'device_type': 'cisco_nxos',
    'host': '10.0.6.6',
    'password': 'bogus123',
    'username': 'user'
}
{
    'device_type': 'cisco_ios',
    'host': '10.0.5.5',
    'password': 'bogus123',
    'username': 'user'
}
```
